### ISSUE TITLE
sticky: fixed/signed legacy addons not on AMO: No Resource URI Leak etc

### ISSUE THREAD

**Issue by earthlng** [ghacksuserjs/ghacks-user.js#191]
AMO no longer allows the uploading of legacy addons. Unless you are the original developer, you cannot even upload a fixed clone. This is an issue - ESR will be on extended life support for almost a year (until 2018-06-26 I believe).

:small_orange_diamond:  Fixed add-ons

- [NoRedirect (clone)](https://raw.githubusercontent.com/earthlng/testpages/master/noredirect_clone-1.3.2.13.1-fx%2Bsm.xpi) - original AMO [here](https://addons.mozilla.org/firefox/addon/noredirect/)
   * author: earthling
   * fixes the missing `multiprocessCompatible` flag
- [No Resource URI Leak (clone)](https://raw.githubusercontent.com/earthlng/testpages/master/no_resource_uri_leak-1.1.1-an%2Bfx%2Bsm%2Btb.xpi) - original AMO [here](https://addons.mozilla.org/firefox/addon/no-resource-uri-leak/)
   * author: earthling
   * The original add-on broke in FF55+

---
@earthlng says:

FF55 seems to break SDK addons that use a relative path to load content scripts.
Using the absolute path fixes the problem in FF55+ and also works in older versions.

Until the addon devs release an update or someone @ mozilla fixes whatever caused the addon to break, you can use this copy - fixed by Yours truly xD

**You need to configure your settings again** because I had to change the addon ID.

I also removed the console output for caught exceptions (annoying and useless for end-users) + changed the name to make it easier to distinguish.
You can keep the original installed (**disabled!**) and wait for an official update.

[download](https://raw.githubusercontent.com/earthlng/testpages/master/no_resource_uri_leak-1.1.1-an%2Bfx%2Bsm%2Btb.xpi)

Resource Test Page: https://browserleaks.com/firefox

**Comment by rebrovov** [comment]
AWESOME !!!

**Comment by Thorin-Oakenpants** [comment]
Added info to the wiki

**Comment by Thorin-Oakenpants** [comment]
@kaliostro2 - you were waiting for this fix (no resource URI leak) - send some free sweeeeeet sweeeeeet love to earthlng. This message will self destruct in 5 seconds

**Comment by publicarray** [comment]
[Smart Referer](https://addons.mozilla.org/en-US/firefox/addon/smart-referer/) is now a hybrid extension (to migrate preferences). The code for the pure web extension is already on GitHub and is planned to ship in the [beginning of September](https://github.com/meh/smart-referer/pull/48#issuecomment-313885047)

**Comment by AdKiller** [comment]
Thanks for fixing No Resource URI Leak (clone).

**Comment by Thorin-Oakenpants** [comment]
I posted a private blog about this on ghacks and tried to nudge Martin into doing an article. Image the thousands of people wanting this to work in 55. I'll give him another nudge via email

**Comment by Thorin-Oakenpants** [comment]
https://www.ghacks.net/2017/08/09/about-abandoned-firefox-add-ons-breaking/

**Comment by earthlng** [comment]
Great, I didn't read nor include the COPYING file. I hope I can count on you with my legal fees.

**Comment by zymase** [comment]
@earthlng added 

> fixed by Yours truly xD

and the audience replied : Bravo, thank you, merci, danke schön :)

This sticky topic is and will become a reference for FF55+ users. Already mentioned on Ghacks.

> Using the absolute path fixes the problem in FF55+ and also works in older versions.

Here on FF ESR 52 I have no issue with the original _No Resource URI Leak 1.1.0_ but is it advised to nevertheless update this add-on to its 1.1.1 version proposed here?

Many thanks, earthlng.

**Comment by Thorin-Oakenpants** [comment]
Sure, send the bill to Atavic .. I have promoted him to Secretary of Finance

**Comment by earthlng** [comment]
> Here on FF ESR 52 I have no issue with the original No Resource URI Leak 1.1.0 but is it advised to nevertheless update this add-on to its 1.1.1 version proposed here?

No. If the console output for caught exceptions in the original version bother you then you can use my version instead. Apart from that they work identical.

**Comment by zymase** [comment]
I've nevertheless updated _No Resource URI Leak 1.1.0_ to _No Resource URI Leak (clone) 1.1.1_ even if not required with pre-FF55 for the sake of code improvement. "_It's not because it works that you shouldn't improve it_"

**Comment by rekixex** [comment]
@Thorin-Oakenpants I am curious, is Resource URI Leak still a problem on Firefox 55.0 with common en-US locale and not many installed add-ons (about 10)? Can websites _enumerate_ installed web extensions / legacy add-ons, or even access add-on _preferences_?

**Comment by rekixex** [comment]
@Thorin-Oakenpants I am curious, is Resource URI Leak still a problem on Firefox 55 with common en-US locale and not many installed add-ons (about 10)? Can websites _enumerate_ installed web extensions / legacy add-ons, or even access add-on _preferences_?

**Comment by rekixex** [comment]
@Thorin-Oakenpants I am curious, is Resource URI Leak still a problem on Firefox 55 with common en-US locale and not many installed add-ons (about 10)? Can websites _enumerate_ installed legacy add-ons / web extensions, or even access add-on _preferences_?

**Comment by rekixex** [comment]
@Thorin-Oakenpants I am curious, is Resource URI Leak still a problem on Firefox 55 with common en-US locale and not many installed add-ons (about 10)? Can websites enumerate installed legacy add-ons / web extensions, or even _access add-on preferences_?

**Comment by publicarray** [comment]
@rekixex You can have a look at https://browserleaks.com/firefox it shows you what is leaked. I don't think the locale matters in any way.

> Can websites enumerate installed legacy add-ons / web extensions, or even access add-on preferences?

Checkout https://thehackerblog.com/addon_scanner. 

More [fingerprinting and other test sites are in the wiki](https://github.com/ghacksuserjs/ghacks-user.js/wiki/Appendix-C:-Test-Sites)

P.S. Could you please stop spamming my email? Much appreciated.

**Comment by Thorin-Oakenpants** [comment]
> [browserleaks] it shows you what is leaked

^^ TBH, this shows you a few things that can be leaked, focusing on FF/TBB internals

Resource:// URIs declared `content-accessible` in manifests LEAK even with the proposed tor uplift patch (otherwise shit breaks), in fact they are _"Moving resources that **need** to be exposed to web content to locations that are marked as content-accessible"_  - this so that ALL OTHER resources are off limits. This simplifies it. I am not a dev, I do not think this manifest is required/used in Web Ext.

There are other ways to detect addons - notably Adblockers, although TBH, a lot of them are generic and do not actually detect the extension, just the affect, AFAIK. I am sometimes accused of having AdBlock Plus, which I don't.

Preferences can't be accessed directly, only [induced](https://en.wikipedia.org/wiki/Inductive_reasoning) - if they could be directly read, we'd be FP'able to the nth degree! [Note: this is what Sherlock Holmes almost always does - not [deduction](https://en.wikipedia.org/wiki/Deductive_reasoning) as he claims so often - _tidbit included for crssi_ to help my beer funds ]

**Comment by crssi** [comment]
:)

Unfortunately I have understood it all. But to fill your jar, I can easily find some stupid question for beer sake?

**Comment by earthlng** [comment]
> I do not think this manifest is required/used in Web Ext.

It is used in WE's, think of Decentraleyes for example which needs to be able to make the scripts it ships with available to the content

> Preferences can't be accessed directly, only induced

all the default preferences files in FF can be accessed and read directly.

**Comment by Thorin-Oakenpants** [comment]
> atm all the default preferences files in FF can be accessed and read directly.

Well, then that creates what? 10 distinct default pref sets? - Mac, Windows, Mobile, Focus, Various Linux? I can see default prefs giving away OS, build even. So loads of info just on defaults. OS, versions, builds... wouldn't be hard to work out combos unique to each one.

Out of curiosity .. how is this accessible _ATM_ ? resource leak?

**Comment by earthlng** [comment]
disable the addon and load https://browserleaks.com/firefox, then click on any of the filenames it detected and it will list you all the prefs in that file. OS and build can easily be detected that way if you look for only a handful of prefs. Check for a single pref that's **new** in 55 or a single pref that has a unique value in Mac or Linux, etc.

**Comment by rekixex** [comment]
I know there may be other ways to detect add-ons, I was curious about add-on detection via resource://. If this leak reveals only default platform specific preferences, I don't see why it is considered problematic - when you don't fake your browser locale or OS websites can already detect it even without resource:// URI leak.

On https://browserleaks.com/firefox it is stated that:

> Mozilla developers is not consider the resources as a fingerprinting vector, despite the fact that some of them can reveal what the user does not wish. For example, differences in built-in preferences files clearly indicates you are using Windows or Linux or Mac, even if you're behind Tor Browser.

On the other hand, No Resource URI Leak developers say:

> Web sites will look for installed extensions on the user's browser. This is terrible for privacy.

> Especially, if you have an ad blocker, or a privacy add-on, or many add-ons installed, you are really strongly encouraged to try this add-on to protect your privacy.

https://discourse.mozilla.org/t/support-no-resource-uri-leak/9004

**Comment by earthlng** [comment]
> I don't see why it is considered problematic

Did you see @publicarray's [comment](https://github.com/ghacksuserjs/ghacks-user.js/issues/191#issuecomment-323008726)? Addons can be detected via resource:// as well

**Comment by Thorin-Oakenpants** [comment]
> f.e. check for a single pref that's **new** in 55

It needs to be more precise that than, because that pref would only indicate FF55-PLUS. Like I said it wouldn't be hard to find combos of prefs+values unique to Every. Single. FF. Build. And. OS.

I would rather be in the small group (until 57 hopefully) that creates a data bit for FP of "access denied"

**Comment by L-a-n-g-o-l-i-e-r-s** [comment]
Was this issue resolved with today's update? :8ball:

**Comment by Thorin-Oakenpants** [comment]
@L-a-n-g-o-l-i-e-r-s - what issue? Do you mean resource URI - the ticket is https://bugzilla.mozilla.org/show_bug.cgi?id=863246 - unresolved as yet. If anything happens, it still takes time - it has to go thru NIghtly, then Beta, then land in stable. So 57 would be the earliest. Dot releases do not contain new features.

If you mean the addon, nah - if a legacy addon breaks it's because legacy code has been stripped - no coming back from that

**Comment by L-a-n-g-o-l-i-e-r-s** [comment]
@Thorin-Oakenpants 
August 25, 2017
Version 55.0.3, first offered to Release channel users on August 25, 2017
    Fix an issue with addons when using a path containing non-ascii characters (bug 1389160)

This didn't fix that issue? I thought the point was to mark everything legacy then tear the legs out from under it in 57. They're nothing if not inconsistent. :rage: :coffin: 

Thanks :sweat_smile:

**Comment by 2glops** [comment]
@L-a-n-g-o-l-i-e-r-s 
I think that fix the issue with the original addon internal code itself. The original addon should now works as well as the clone addon.
But FF 53.0.3 don't fix the issue of accessing resource:// access to Web content.

**Comment by L-a-n-g-o-l-i-e-r-s** [comment]
@2glops 
Right, I know the URI issue hasn't been fixed, but I was thinking that the update should have fixed the extension among the others that broke, is that true? :disappointed: 

Thanks

**Comment by earthlng** [comment]
> I was thinking that the update should have fixed the extension among the others that broke, is that true?

No that's not true. That bug in the update fixed an unrelated problem:

https://bugzilla.mozilla.org/show_bug.cgi?id=1389160#c47

> [User impact if declined]: users with non-ascii in path (such as unicode in username) will start up with add-ons disabled

**Comment by Thorin-Oakenpants** [comment]
@earthlng 
> > Preferences can't be accessed directly, only induced
> 
> atm all the default preferences files in FF can be accessed and read directly.

For some reason this question was in my head when I woke up: Hey sure, it can "read" what files are there, but surely it cannot "upload" them from your device and inspect the contents, right? That doesn't sound right to me. Or do you mean the local file can be integrated into content and read by JS? Can someone ELI5 how a website would read the actual pref values

**Comment by earthlng** [comment]
> Can someone ELI5 how a website would read the actual pref values

I already did ... https://github.com/ghacksuserjs/ghacks-user.js/issues/191#issuecomment-323073448

They cannot read it in it's raw form including the js comments for example, but they can re-construct the active code parts because there are only pref() and sticky_pref() functions in those files. They could then send the reconstructed file to a server. But there's no need to do that because a handful of those pref calls are enough to detect the OS + FF version.
The browserleaks site even shows you example code that illustrates how it works.

**Comment by Thorin-Oakenpants** [comment]
^^ Just as well its all locked down then. I feel a good nights sleep coming on .. no wait a minute ... damn, did you see my latest soapbox post?

**Comment by Atavic** [comment]
Secretary of Finance? I don't even own a credit card. :feelsgood:

**Comment by fmarier** [comment]
From <https://groups.google.com/d/msg/mozilla.dev.platform/00-1tT15mX0/TzUrOD93AAAJ>:

> It means there is no need to use the "No Resource URI Leak" add-on anymore.

**Comment by ilikenwf** [comment]
Yes, unless you use an ESR release, or Waterfox until it's merged in. That said someone here should test it first to see if it really passes the tests.

My modded version also should prevent extension fingerprinting though.

**Comment by earthlng** [comment]
Hi @ilikenwf 

thanks. You'll not gonna be able to get it signed with the same id but more importantly I think you're breaking the logic here:
```
-  || u.schemeIs ('about') && (!policyState.veryInsecureAboutUris.has (u.path))
-    && (policyState.secureAboutUris.has (u.path) || policyState.whitelistAboutUris);
+  || u.schemeIs ('about') || u.schemeIs ('extension') || u.schemeIs ('moz-extension') 
+	&& (!policyState.veryInsecureAboutUris.has (u.path))
+    && (policyState.secureAboutUris.has (u.path) || policyState.mozextWhitelist.has (u.path) 
+	|| policyState.extWhitelist.has (u.path) || policyState.whitelistAboutUris);
```
I'll have to look at it in more detail, tomorrow maybe.

**Comment by ilikenwf** [comment]
Waterfox doesn't force signing, so I wasn't super worried about it since I've put in a PR to the addon dev's repo...

I don't think the logic there is broken, necessarily, but it may need some parenthases for clarity because it's hard to read and was so before I even touched it...but the logic reads to me as "if the scheme is any of the things we filter and our policy of insercure about uris does not have the path, and the secure uris, or other policies has the path.....

**Comment by earthlng** [comment]
> I don't think the logic there is broken

No I'm pretty sure it's broken. It needs to be **||** `u.schemeIs ('about')` **&&** `all the rest of it`.
You made it **||** `u.schemeIs ('moz-extension')` **&&** `all the rest of it`.

> it's hard to read and was so before I even touched it

I know, I nearly broke it myself when I tried to add some improvements to the addon  ;)

**Comment by ilikenwf** [comment]
Pushed that tweak.

**Comment by ilikenwf** [comment]
Yeah, still broken, that whitelist logic needs adjusted...the blocking works but the whitelists do not, meaning ublock etc options don't look or function properly.

**Comment by Thorin-Oakenpants** [comment]
These two patched legacy extensions are listed on the wiki page. I doubt anything else will ever be patched by us - closing this
